File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:

        hashMap = {}
        n1 = len(nums1)
        n2 = len(nums2)

        for index, num in enumerate(nums2):
            hashMap[num] = index

        stack = []

        for i in reversed(range(n2)):
            if not stack:
                stack.append(nums2[i])
                nums2[i] = -1
            else:
                while stack and stack[-1] < nums2[i]:
                    stack.pop()
                if not stack:
                    stack.append(nums2[i])
                    nums2[i] = -1
                else:
                    ele = nums2[i]
                    nums2[i] = stack[-1]
                    stack.append(ele)
        
        for index, num in enumerate(nums1):
            nums1[index] = nums2[hashMap[num]]

        return nums1
        

File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {

        int n1 = nums1.length;
        int n2 = nums2.length;

        Map<Integer, Integer> hashMap = new HashMap<>();

        for (int i = 0; i < n2; i++){
            hashMap.put(nums2[i], i);
        }

        Stack<Integer> stack = new Stack<>();

        for (int i = n2-1; i >= 0; i--) {
            if (stack.isEmpty()) {
                stack.push(nums2[i]);
                nums2[i] = -1;
            } else {
                while (!stack.isEmpty() && stack.peek() <= nums2[i]){
                    stack.pop();
                }
                if (stack.isEmpty()) {
                    stack.push(nums2[i]);
                    nums2[i] = -1;
                } else {
                    int ele = nums2[i];
                    nums2[i] = stack.peek();
                    stack.push(ele);
                }
            }
        }

        for (int i = 0; i < n1; i++) {
            int ele = nums1[i];
            nums1[i] = nums2[hashMap.get(ele)];
        }

        return nums1;
        
    }
}